feat(study-search): optimize the studies search engine

mabw-rte opened the pull request:
context: 
- users have noticed a slow processing when searching for studies

issue:
- the filtering need to be executed at the database level

solution:
- include all filters for the database studies query

mabw-rte committed:
feat(study-search): optimize the studies search engine

laurent-laporte-pro commented on antarest/study/repository.py:
```diff
@@ -1,18 +1,96 @@
 
+def escape_like(string: str, escape_char: str = "\\") -> str:
+    """
+    Escape the string parameter used in SQL LIKE expressions.
```
Il faudrait utiliser le format à la Google Doc au lieu de Sphinx :-)

laurent-laporte-pro commented on antarest/study/repository.py:
```diff
@@ -101,30 +179,73 @@ def get_additional_data(self, study_id: str) -> t.Optional[StudyAdditionalData]:
+        pagination: StudyPagination = StudyPagination(),
     ) -> t.List[Study]:
+        """
+        This function goal is to create a search engine throughout the studies with optimal
```
Je préfère que l'on reformule en essayant de répondre à la question "Que fait la fonction get_all ?": "La fonction permet de rechercher…". Merci de documenter les paramètres et la valeur retour. Un exemple d'utilisation pourrait être utile.

laurent-laporte-pro commented on antarest/study/repository.py:
```diff
@@ -101,30 +179,73 @@ def get_additional_data(self, study_id: str) -> t.Optional[StudyAdditionalData]:
 
         q = self.session.query(entity)
-        if exists:
+        if study_filter.exists:
```
Il y a 3 cas à considérer :  `None`, `True` et `False`

laurent-laporte-pro commented on antarest/study/repository.py:
```diff
@@ -101,30 +179,73 @@ def get_additional_data(self, study_id: str) -> t.Optional[StudyAdditionalData]:
                 q = q.filter(RawStudy.workspace != DEFAULT_WORKSPACE_NAME)
-        if studies_ids is not None:
-            q = q.filter(entity.id.in_(studies_ids))
+        if study_filter.studies_ids is not None:
```
En fait, j'aime pas trop le `None`. Il faudrait plutôt examiner les cas où on appelle la fonction `get_all` avec une liste vide. C'est plutôt là qu'il faudrait corriger.

laurent-laporte-pro commented on antarest/study/repository.py:
```diff
@@ -101,30 +179,73 @@ def get_additional_data(self, study_id: str) -> t.Optional[StudyAdditionalData]:
+            regex = f"{escape_like(study_filter.folder)}%"
+            q = q.filter(entity.folder.ilike(regex))
+        if study_filter.workspace:
+            regex = f"%{escape_like(study_filter.workspace)}%"
```
Pas de regex ici. On veut une recherche exacte.

laurent-laporte-pro commented on antarest/study/repository.py:
```diff
@@ -101,30 +179,73 @@ def get_additional_data(self, study_id: str) -> t.Optional[StudyAdditionalData]:
+            q = q.order_by(entity.name.desc())
+        elif sort_by == StudySortBy.NAME_ASC:
+            q = q.order_by(entity.name.asc())
+
```
Ajoute :

```python
else:  # pragma: no cover
    raise NotImplementedError(sort_by)
```

laurent-laporte-pro commented on antarest/study/service.py:
```diff
@@ -438,44 +439,33 @@ def edit_comments(
-        workspace: Optional[str],
-        folder: Optional[str],
         params: RequestParameters,
+        study_filter: StudyFilter,
```
Pourquoi `study_filter` est un paramètre obligatoire et les autres sont optionnels ?

mabw-rte committed:
feat(study-search): improve doc, code review

mabw-rte committed:
test(study-search): add unittests for the filtering parameters

mabw-rte committed:
test(study-search): add integration tests for the studies listing

laurent-laporte-pro committed:
docs(study-search): correct docstring in `StudyFilter` and `StudyPagination`

laurent-laporte-pro committed:
fix(study-search): correct get_studies endpoint and add unit tests with error checking

laurent-laporte-pro committed:
style(study-search): correct typo in comment

laurent-laporte-pro committed:
feat(study-search): turn the `sortBy` parameter into an optional parameter

It is advisable to use an optional Query parameter for enumerated types, like booleans.

laurent-laporte-pro committed:
style(study-search): use `t` alias to import `typing`

laurent-laporte-pro committed:
chore(study-search): add a todo to use `with_polymorphic`

I think we should use a `entity = with_polymorphic(Study, "*")` to make sure RawStudy and VariantStudy fields are also fetched.

laurent-laporte-pro committed:
feat(study-search): improve the get_studies endpoint to better check `versions` and `users` parameters

laurent-laporte-pro committed:
fix(study-search): avoid polynomial runtime due to backtracking in regex (SonarCloud issue)

laurent-laporte-pro committed:
style(study-search): correct issue with mypy linter
